[tests] Refactor MongoDB.Driver tests to use testcontainers

radical opened the pull request:
###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2718)

radical committed:
[tests] Refactor MongoDB.Driver tests to use testcontainers

radical committed:
Run MongoDB.Driver tests on helix

radical committed:
RequiredDocker: disable for windows on CI - build machine, and helix

radical committed:
fix condition

radical committed:
Fix the health tests when container is not available

radical committed:
Use a property for the nuget version

eerhardt commented on eng/Versions.props:
```diff
@@ -46,5 +46,6 @@
     <MicrosoftEntityFrameworkCoreSqlServerPackageVersion>8.0.2</MicrosoftEntityFrameworkCoreSqlServerPackageVersion>
     <MicrosoftEntityFrameworkCoreToolsPackageVersion>8.0.2</MicrosoftEntityFrameworkCoreToolsPackageVersion>
     <MicrosoftNETRuntimeWorkloadTestingInternalVersion>9.0.0-preview.3.24151.4</MicrosoftNETRuntimeWorkloadTestingInternalVersion>
+    <TestcontainersPackageVersion>3.7.0</TestcontainersPackageVersion>
```
Let's leave `eng\Versions.props` only have the versions that are managed by darc in it. Can you move this shared property to the top of `Directory.Packages.props`? I dislike having to look in multiple places. And when we get dependabot working again, I'm not sure it knows how to update `eng\Versions.props` files.

eerhardt commented on tests/Shared/RequiresDockerTheoryAttribute.cs:
```diff
@@ -3,12 +3,15 @@
+    // Not available on windows/CI
+    public static bool IsSupported => !OperatingSystem.IsWindows() ||
+                                      (Environment.GetEnvironmentVariable("HELIX_WORKITEM_ROOT") is null
+                                      && Environment.GetEnvironmentVariable("BUILD_BUILDID") is null);
```
Why do we need to add `BUILD_BUILDID`? 

I assume you are trying to support the case where these tests try to run on the azdo build pipeline machines, but does Docker work on the azdo pipeline machines on Linux?

radical replied:
This was to make sure that we don't run any `RequiresDocker*` tests on windows/CI at all - helix or build machines.

eerhardt commented on tests/Aspire.RabbitMQ.Client.Tests/Aspire.RabbitMQ.Client.Tests.csproj:
```diff
@@ -6,6 +6,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <Compile Include="..\Shared\RequiresDockerTheoryAttribute.cs" />
```
We have a shared test library. Let's move this there instead so all these projects don't need to compile include the file. You can also put it in an existing namespace like `Aspire` or `Aspire.Components`, so all the files don't need to include a new one-off namespace.

radical replied:
Where is this shared test library? I don't see a csproj in `tests/Shared`.

eerhardt replied:
https://github.com/dotnet/aspire/tree/main/tests/Aspire.Components.Common.Tests

eerhardt commented on tests/Aspire.MongoDB.Driver.Tests/ConformanceTests.cs:
```diff
@@ -92,6 +104,11 @@ protected override void TriggerActivity(IMongoClient service)
         service.ListDatabases(source.Token);
     }
 
+    protected override HealthStatus GetHealthStatus()
+        => _containerFixture.Container?.Health == TestcontainersHealthStatus.Healthy
+            ? HealthStatus.Healthy
+            : HealthStatus.Unhealthy;
```
We should always be expecting that we can connect to the server now and that it is healthy. We shouldn't need this new virtual/override method. We just need to `override bool CanConnectToServer => RequiresDockerTheoryAttribute.IsSupported;`, which will make the `HealthCheckReportsExpectedStatus` test work correctly.

radical replied:
It was failing because the it got `Unhealthy` from the service.

https://github.com/radical/aspire/blob/46bce9b06b4e6ea10e3a68879bad53d24cb9c2a8/tests/Aspire.Components.Common.Tests/ConformanceTests.cs#L317-L320

eerhardt replied:
If the service says it is unhealthy, then the test should fail. We don't want to mask that problem. We want to know when something goes wrong.

radical replied:
Right, and this would get the real status of the container to match against what we get from the health check. I'll try to see why/when it is showing unhealthy.

eerhardt commented on tests/Aspire.MongoDB.Driver.Tests/AspireMongoDBDriverExtensionsTests.cs:
```diff
@@ -10,13 +11,21 @@
+
     [Theory]
-    [InlineData("mongodb://localhost:27017/mydatabase", true)]
+    [InlineData("mongodb://localhost:2017/mydatabase", true)]
```
Why this change?

radical replied:
debugging slipped through!

radical committed:
move package version property to Directory.Packages.props

radical committed:
fix typo

radical committed:
address review feedback

radical committed:
Move RequiresDockerTheoryAttribute.cs to the Common project - addresses review feedback

radical committed:
Remove /test_db from the connection string, so it defaults to using admin db to auth

- And explicitly use mongo:7.0.5

TODO: pick up the version to match what we use for aspire.hosting

radical commented:
I'm not sure if removing `test_db` is the correct thing. Or if need to prime the db after creation, and before the tests are run.
 [`308535c` (#2718)](https://github.com/dotnet/aspire/pull/2718/commits/308535cdc28cc86ed77838e8189d8765a54be47c)

radical committed:
Create a test_db with user mongo, pwd:mongo as expected by the tests

radical committed:
Merge remote-tracking branch 'origin/main' into mongodb-testcontainers

radical committed:
Merge remote-tracking branch 'origin/main' into mongodb-testcontainers

eerhardt commented on tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:
```diff
@@ -0,0 +1,49 @@
+            await Container.StartAsync();
+
+            // Create `test_db` database with user:mongo pwd:mongo
+            var mongoClient = new MongoClient(Container.GetConnectionString());
+            var createUserCommand = new BsonDocumentCommand<BsonDocument>(BsonDocument.Parse("""
+            {
+               createUser: "mongo",
+               pwd: "mongo",
+               roles: [ { role: 'readWrite', db: 'test_db' } ]
+            }
+            """));
+            await mongoClient.GetDatabase("test_db")
+                .RunCommandAsync(createUserCommand);
```
In our "playground" app, we just do:

https://github.com/dotnet/aspire/blob/ded1eaef1592000b7c99136d1cc9a9085170f07b/playground/mongo/Mongo.ApiService/Program.cs#L19-L20

Is that enough? That seems simpler than what we have here.

radical replied:
With the only the two lines above, it still fails with:
`,"msg":"Failed to authenticate","attr":{"client":"172.17.0.1:58864","isSpeculative":false,"isClusterMember":false,"mechanism":"SCRAM-SHA-1","user":"mongo","db":"test_db","error":"UserNotFound: Could not find user \"mongo\" for db \"test_db\"",`

The testcontainers use `mongo:mongo` by default.
How is the container setup/started in case of the playground?

eerhardt replied:
> How is the container setup/started in case of the playground?

https://github.com/dotnet/aspire/blob/ded1eaef1592000b7c99136d1cc9a9085170f07b/playground/mongo/Mongo.AppHost/Program.cs#L6

radical replied:
If the default is to start it without any user/pass then we can do the same in tests too.
```csharp
            Container = new MongoDbBuilder()
                .WithImage("mongo:7.0.5")
                .WithUsername(null)
                .WithPassword(null)
                .Build();
```

eerhardt commented on tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:
```diff
@@ -0,0 +1,49 @@
+        if (RequiresDockerTheoryAttribute.IsSupported)
+        {
+            Container = new MongoDbBuilder()
+                .WithImage("mongo:7.0.5")
```
I wonder if we should make it so these versions stay in sync between our tests and the AppHost:

https://github.com/dotnet/aspire/blob/ded1eaef1592000b7c99136d1cc9a9085170f07b/src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs#L31

radical replied:
Yes, definitely!

radical replied:
Do the versions need to be surfaced in the api too? If it is just for tests, and wherever they are used in the repo right now, then we can use a shared `.cs` file for that.

eerhardt replied:
> Do the versions need to be surfaced in the api too?

No, I don't believe so.

> we can use a shared .cs file for that.

That sounds good to me.

cc @mitchdenny

radical replied:
I'll cover this in a follow up PR.

radical committed:
mongodb: don't use auth in the tests

radical committed:
Merge remote-tracking branch 'origin/main' into mongodb-testcontainers

radical committed:
fix build

eerhardt reviewed (approved):
Nice! This looks really clean. LGTM
